packages/components/src/shortcuts/Shortcut.ts

```diff
@@ -0,0 +1,324 @@
+/* eslint-disable no-underscore-dangle */
+import React from 'react';
+import { ContextActionUtils } from '../context-actions';
+
+export enum MODIFIER {
+  CTRL = 'MODIFIER_CTRL',
+  CMD = 'MODIFIER_CMD',
+  ALT = 'MODIFIER_ALT',
+  OPTION = 'MODIFIER_OPTION',
+  SHIFT = 'MODIFIER_SHIFT',
+}
+
+// The value should match the KeyboardEvent.key value
+// This is the value which will be displayed if no display overrides are done in the OS specific getDisplayString methods
+export enum KEY {
+  A = 'A',
+  B = 'B',
+  C = 'C',
+  D = 'D',
+  E = 'E',
+  F = 'F',
+  G = 'G',
+  H = 'H',
+  I = 'I',
+  J = 'J',
+  K = 'K',
+  L = 'L',
+  M = 'M',
+  N = 'N',
+  O = 'O',
+  P = 'P',
+  Q = 'Q',
+  R = 'R',
+  S = 'S',
+  T = 'T',
+  U = 'U',
+  V = 'V',
+  W = 'W',
+  X = 'X',
+  Y = 'Y',
+  Z = 'Z',
+  BACKSPACE = 'Backspace',
+  ESCAPE = 'Escape',
+  ENTER = 'Enter',
+  DELETE = 'Delete',
+  SLASH = '/',
+}
+
+type ShortcutKeys = [...MODIFIER[], KEY];
+
+export interface KeyState {
+  /**
+   * This is purposely keyValue and not key to make KeyboardEvents not match the interface
+   * KeyboardEvents need some processing to get the actual value
+   * Use Shortcut.getKeyStateFromEvent to get the right KeyState from an event
+   */
+  keyValue: string | null;
+  altKey: boolean;
+  ctrlKey: boolean;
+  metaKey: boolean;
+  shiftKey: boolean;
+}
+
+interface ValidKeyState extends KeyState {
+  keyValue: KEY;
+}
+
+export default class Shortcut {
+  readonly id: string; // Unique identifier for the shortcut
+
+  readonly name: string; // e.g. Rename, Save, Run Selected
+
+  readonly tooltip?: string;
+
+  readonly isEditable: boolean;
+
+  private readonly defaultKeyState: ValidKeyState;
+
+  private keyState: ValidKeyState;
+
+  /**
+   * Use to check if a keyCode corresponds to an allowed key for a shortcut
+   * @param keyCode The keyCode to check. This should be the charCode of the key
+   * @returns Type predicate asserting the key is an allowed KEY
+   */
+  static isAllowedKey(key: string | null): key is KEY {
+    return Object.values(KEY).includes(key as KEY);
+  }
+
+  /**
+   * Checks if a KeyState has a valid key.
+   * @param state KeyState with any string as the key
+   * @returns True if KeyState is is using an allowed keyCode
+   */
+  static isValidKeyState(state: KeyState): state is ValidKeyState {
```

Comment:
Why are we doing a check like this? What about ESCAPE, BACKSPACE, arrow keys... why don't we let a user set a keyboard shortcut to just a letter and not require a modifier key?